pkg/cmd/server/kubernetes/node.go

```diff
@@ -189,73 +157,76 @@ func (c *NodeConfig) RunKubelet() {
 		DockerFreeDiskMB: 256,
 		RootFreeDiskMB:   256,
 	}
-
-	k, err := kubelet.NewMainKubelet(
-		c.NodeHost,
-		c.DockerClient,
-		c.Client,
-		c.VolumeDir,
-		c.ImageFor("pod"),
-		3*time.Second,
-		0.0,
-		10,
-		gcPolicy,
-		cfg.SeenAllSources,
-		c.ClusterDomain,
-		clusterDNS,
-		kapi.NamespaceDefault,
-		app.ProbeVolumePlugins(),
-		app.ProbeNetworkPlugins(),
-		c.NetworkPluginName,
-		5*time.Minute,
-		recorder,
-		cadvisorInterface,
-		imageGCPolicy,
-		diskSpacePolicy,
-		nil,
-		15*time.Second,
-		"/kubelet",
-		kubecontainer.RealOS{},
-		"",
-		"docker",
-		mount.New(),
-		"", // docker daemon container
-	)
+	kubeAddress, kubePortStr, err := net.SplitHostPort(c.BindAddress)
 	if err != nil {
-		glog.Fatalf("Couldn't run kubelet: %s", err)
+		glog.Fatalf("Cannot parse node address: %v", err)
 	}
-	go util.Forever(func() { k.Run(cfg.Updates()) }, 0)
-
-	handler := kubelet.NewServer(k, true)
-
-	server := &http.Server{
-		Addr:           c.BindAddress,
-		Handler:        &handler,
-		ReadTimeout:    5 * time.Minute,
-		WriteTimeout:   5 * time.Minute,
-		MaxHeaderBytes: 1 << 20,
+	kubePort, err := strconv.Atoi(kubePortStr)
+	if err != nil {
+		glog.Fatalf("Cannot parse node port: %v", err)
 	}
 
-	go util.Forever(func() {
-		glog.Infof("Started Kubelet for node %s, server at %s, tls=%v", c.NodeHost, c.BindAddress, c.TLS)
-		if clusterDNS != nil {
-			glog.Infof("  Kubelet is setting %s as a DNS nameserver for domain %q", clusterDNS, c.ClusterDomain)
-		}
-		k.BirthCry()
-
-		if c.TLS {
-			server.TLSConfig = &tls.Config{
+	var tlsOptions *kubelet.TLSOptions
+	if c.TLS {
+		tlsOptions = &kubelet.TLSOptions{
+			Config: &tls.Config{
 				// Change default from SSLv3 to TLSv1.0 (because of POODLE vulnerability)
 				MinVersion: tls.VersionTLS10,
 				// RequireAndVerifyClientCert lets us limit requests to ones with a valid client certificate
 				ClientAuth: tls.RequireAndVerifyClientCert,
 				ClientCAs:  c.ClientCAs,
-			}
-			glog.Fatal(server.ListenAndServeTLS(c.KubeletCertFile, c.KubeletKeyFile))
-		} else {
-			glog.Fatal(server.ListenAndServe())
+			},
+			CertFile: c.KubeletCertFile,
+			KeyFile:  c.KubeletKeyFile,
 		}
-	}, 0)
+	}
+
+	kcfg := kapp.KubeletConfig{
+		Address: util.IP(net.ParseIP(kubeAddress)),
+		// Allow privileged containers
+		// TODO: make this configurable and not the default https://github.com/openshift/origin/issues/662
+		AllowPrivileged:                true,
+		HostNetworkSources:             hostNetworkCapabilities,
+		HostnameOverride:               c.NodeHost,
+		RootDirectory:                  c.VolumeDir,
+		ConfigFile:                     c.PodManifestPath,
+		ManifestURL:                    "",
+		FileCheckFrequency:             time.Duration(c.PodManifestCheckIntervalSeconds) * time.Second,
+		HTTPCheckFrequency:             0,
+		PodInfraContainerImage:         c.ImageFor("pod"),
+		SyncFrequency:                  3 * time.Second,
```

Comment:
Our sync interval should be whatever upstream uses if it's longer than what we use now.